.github/workflows/cicd.yml

```diff
@@ -178,17 +178,20 @@ jobs:
 
     needs: [lint, mypy, docs, packaging]
 
-    runs-on: ubuntu-latest
+    runs-on: ${{ matrix.os }}
     timeout-minutes: 30
     strategy:
       matrix:
+        os: ["ubuntu-latest"]
         python: ["3.6", "3.7", "3.8", "3.9", "pypy3"]
         twisted: ["19.2", "current"]
 
     steps:
 
       - name: Checkout source code
         uses: actions/checkout@v2
+        with:
+          fetch-depth: 2
```

Comment:
This is needed by the Codecov action.

Reply:
Yes. That should be ok.
For PR builds, you can use the extra depth to get the last commit message... as HEAD is a merge with the PR target branch.

For paid runs I use this to skip a build on macos for example... when I am working on fixing a windows only bug. 

```
    - name: Fail on "skip macos"
      if: ${{ github.event.after }}
      run: git log -1 --pretty=format:"%s" ${{ toJSON(github.event.after) }} | grep -v 'skip macos'
```

Not an issue with klein as GitHub Action is free )